### PR TITLE
PR56: CLI improvement, repository management

### PR DESCRIPTION
**Pull request by chevdor**
This PR adds a bunch of commands to the cli to list and retrieve compilers.
Best summarised from the command line:
[asciicast](https://asciinema.org/a/86834)

### CODE REVIEW

**Comment by chevdor** [comment]
PR #55 should be merged before this one.

**Comment by sebs** [comment]
Please add the knowledge about downloads and respective hashes to the  source code and displayed output.

File: lib/compilers.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    getListLocal: getListLocal,
+    pathExistsIn: pathExistsIn,
+    downloadBinary: downloadBinary,
+
+    // Use this if you want to add wrapper functions around the pure module.
+    compilers: compilers
+  };
+}
+
+module.exports = compilers;
```
Comment (sebs) [inline lib/compilers.js]: 208 loc? Mayabe this deserves any kind of split

File: lib/helpers.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
+var fs = require('fs-extra');
+var chalk = require('chalk');
+
+function helpers () {
+  function extractCompilerData (path) {
+    var res = {};
+    var match = /v\d\.\d\.\d/.exec(path);
```
Comment (sebs) [inline lib/helpers.js]: please comment the regex: waht are you doing

File: lib/helpers.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    if (match) { res.version = match[0]; }
+
+    match = /(nightly.+)\+/.exec(path);
+
+    if (match) {
+      res.prerelease = match[1];
+    }
+
+    match = /(commit\.\w+)\./.exec(path);
```
Comment (sebs) [inline lib/helpers.js]: please comment the regex: waht are you doing. Lots of regex anyway

File: lib/helpers.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    return res;
+  }
+
+  return {
+    extractCompilerData: extractCompilerData,
+
+    isNode: function () {
+      return (typeof (process) !== undefined);
```
Comment (sebs) [inline lib/helpers.js]: is this "the" way to do it?

File: lib/helpers.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    },
+
+    getSolidityCompiler: function (compilerRepo, compiler) {
+      // here we get the default version of the compiler
+      var solc = require('./../index.js');
```
Comment (sebs) [inline lib/helpers.js]: .js?

File: lib/helpers.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+      // if the user needs a specific one, we load it
+      if (compiler) {
```
Comment (sebs) [inline lib/helpers.js]: check existence andy maybe ONE property that makes compiler a compiler

File: package.json [hunk 7]
```diff
@@ -33,7 +35,10 @@
   },
   "homepage": "https://github.com/ethereum/solc-js#readme",
   "dependencies": {
+    "chalk": "^1.1.3",
+    "fs-extra": "^0.30.0",
     "memorystream": "^0.3.1",
+    "pwuid": "^1.1.1",
```
Comment (sebs) [inline package.json]: pwuid is good!

File: solcjs [hunk 8]
```diff
@@ -1,72 +1,326 @@
...
+if (cmd === 'clean') {
+  console.log('I let you do this one. You compilers are in:', compilers.repository);
+  helpers.exit();
+}
 
-for (var contractName in output.contracts) {
-  if (argv.bin) {
-    fs.writeFileSync(path.join(destination, contractName + '.bin'), output.contracts[contractName].bytecode);
+// Usage:
+// solcjs get             => fetch latest release unless already there
```
Comment (sebs) [inline solcjs]: remove any out commented code?

File: solcjs [hunk 9]
```diff
@@ -1,72 +1,326 @@
...
+    if (requestedVersion) {
+      console.log('Requested version: ' + requestedVersion);
+      wanted = requestedVersion;
+    } else {
+      var pkgVersion = pkg.version.match(/^(\d+\.\d+\.\d+)$/)[1];
+      console.log('Requested version: latest release (' + pkgVersion + ')');
+      wanted = list.getRelease(pkgVersion).path;
+    }
+
+    compilers.downloadBinary(wanted, argv.force, function (file) {
```
Comment (sebs) [inline solcjs]: this should only succeed if the compare to a local stored hash succeeds

File: wrapper.js [hunk 10]
```diff
@@ -85,8 +86,9 @@ function setupMethods (soljson) {
     supportsMulti: compileJSONMulti !== null,
     supportsImportCallback: compileJSONCallback !== null,
     // Use the given version if available.
-    useVersion: function (versionString) {
-      return setupMethods(require('./bin/soljson-' + versionString + '.js'));
+    useVersion: function (compilerRepo, versionString) {
+      versionString = versionString.replace('soljson-', '');
```
Comment (sebs) [inline wrapper.js]: no check for success? replaces silently can fail

File: lib/compilers.js (hunk 1 above)
Comment (chevdor) [inline lib/compilers.js]: I don´t disagree but the current PR is already huge. I am fine taking the time to split if required once the PR is merged.

File: lib/helpers.js (hunk 2 above)
Comment (chevdor) [inline lib/helpers.js]: Regex were not commented in the past I think and this is really basic. If /v\d\.\d\.\d/ is not clear, people will probably not get to the regex anyway :) If more folks want comments there, please mention it, I also don´t mind.

File: wrapper.js (hunk 10 above)
Comment (sebs) [inline wrapper.js]: I am just a bystander and comments reached far more than just the changes on hand for that feature. ;)

File: lib/helpers.js (hunk 4 above)
Comment (chevdor) [inline lib/helpers.js]: What do you propose?

File: solcjs (hunk 8 above)
Comment (chevdor) [inline solcjs]: This is documentation, not commented out code.

File: wrapper.js (hunk 10 above)
Comment (chevdor) [inline wrapper.js]: Can you provide an example of such failure?

**Comment by axic** [comment]
@chevdor this is awesome work! I really like the screencap :)

You have done a lot of work and there are a lot of commits. It is quite hard to review. Would you have time splitting it up in smaller feature merge requests?

**Comment by chevdor** [comment]
Closing in favour of a cleaned up version. See #54

**Pull request closed without merging**